pkg/autodiscovery/listeners/types.go

```diff
@@ -23,19 +24,19 @@ type ContainerPort struct {
 // It should be matched with a check template by the ConfigResolver using the
 // ADIdentifiers field.
 type Service interface {
-	GetEntity() string                         // unique entity name
-	GetTaggerEntity() string                   // tagger entity name
-	GetADIdentifiers() ([]string, error)       // identifiers on which templates will be matched
-	GetHosts() (map[string]string, error)      // network --> IP address
-	GetPorts() ([]ContainerPort, error)        // network ports
-	GetTags() ([]string, string, error)        // tags and tags hash
-	GetPid() (int, error)                      // process identifier
-	GetHostname() (string, error)              // hostname.domainname for the entity
-	GetCreationTime() integration.CreationTime // created before or after the agent start
-	IsReady() bool                             // is the service ready
-	GetCheckNames() []string                   // slice of check names defined in kubernetes annotations or docker labels
-	HasFilter(containers.FilterType) bool      // whether the service is excluded by metrics or logs exclusion config
-	GetExtraConfig([]byte) ([]byte, error)     // Extra configuration values
+	GetEntity() string                                   // unique entity name
+	GetTaggerEntity() string                             // tagger entity name
+	GetADIdentifiers(context.Context) ([]string, error)  // identifiers on which templates will be matched
+	GetHosts(context.Context) (map[string]string, error) // network --> IP address
+	GetPorts(context.Context) ([]ContainerPort, error)   // network ports
+	GetTags() ([]string, string, error)                  // tags and tags hash
+	GetPid(context.Context) (int, error)                 // process identifier
+	GetHostname(context.Context) (string, error)         // hostname.domainname for the entity
+	GetCreationTime() integration.CreationTime           // created before or after the agent start
+	IsReady(context.Context) bool                        // is the service ready
+	GetCheckNames(context.Context) []string              // slice of check names defined in kubernetes annotations or docker labels
+	HasFilter(containers.FilterType) bool                // whether the service is excluded by metrics or logs exclusion config
+	GetExtraConfig([]byte) ([]byte, error)               // Extra configuration values
```

Comment:
As most of the implementations of `Service` do not use `context.Context`, I am wondering if it would be possible to remove `context.Context` and replace it by a `Cancel` method. The `Cancel` method could cancel the last request if any. You can also propagate the information with a channel when a request timeout if you also need this information.


Note: You can create 2 interfaces to avoid a service to implement a empty cancel method when the requests are not cancellable :
```go
type Service interface {
	GetEntity() string
}

type CancellableService interface {
	Service
	Cancel()
}
```
When you have an instance of `Service` you can dynamically check whether the instance also implement `CancellableService` and call `Cancel` only when needed.